File: clients/python/clausters/play.py
"""The free-standing ``play`` — one verb for everything playable.

`play` is the interactive front door: it plays whatever you hand it against the
ambient context, so you never spell out a server or a clock for a quick take.
Like SuperCollider's ``play`` (and sc3's), it dispatches by kind:

- an `clausters.seq.event.Event` — or a plain **dict** of event keys — -> a
  note (immediate outside a clock, timetagged inside one);
- an event `clausters.seq.pattern.Pattern` (a ``Pbind``) -> an
  `clausters.seq.eventstream.EventStreamPlayer` on a clock;
- a `clausters.base.stream.Routine` / `clausters.base.stream.Stream` — or a
  bare **generator** (object or function) -> scheduled on a clock;
- a bare **expression** (a `clausters.defs.Ugen` graph, a
  `clausters.defs.ChannelList` of them, a Faust `clausters.defs.Signal` or
  `clausters.defs.Box`) or a **def**
  (`clausters.defs.SynthDef` / `clausters.defs.FaustDef` /
  `clausters.defs.GraphDef`) -> sent and instanced on the server; an
  expression is first wrapped in an ephemeral def
  (`clausters.defs.asdef.as_def` adds the ``out`` when it lacks one), so
  ``play(sine(440))`` just sounds and ``play(sine(440).dup())`` sounds in
  stereo. Returns the node handle — it plays until you free it;
- a `clausters.seq.timeline.Timeline` -> a `clausters.seq.timeline.Playhead`
  over the ambient clock and server;
- a `clausters.defs.Buffer` -> sounded through the stock playbuf instrument
  (a buffer sounds through an instrument; here the verb provides the default
  one — ``rate``/``amp`` controls, freed when the take ends);
- an `clausters.seq.automation.Automation` -> prepared if needed and
  triggered on the ambient server — the interactive "apply this curve to
  that node's control, now" (outside a clock its beats read as seconds);
- anything else following the **timeline-item protocol**
  (``play(destination)`` — an `OscEvent`, a `MidiEvent`, ...) -> dispatched
  to it with the ambient server.

An arrangement `Element` is **not** playable — its change of state to sound
is `clausters.form.render`.

Everything resolves against the ambient environment (the running session, else
the default session `clausters.default_session`): ``server`` defaults to the
booted default server and ``clock`` to the running routine's clock or, outside
one, the default session's clock (created and started on first use). So after a
single ``Server.boot()``:

```python
from clausters import play
from clausters.seq.event import Event

play(Event(degree=0))                       # one note, now
play(Pbind(degree=Pseq([0, 2, 4]), dur=.5)) # a phrase, on the default clock
```

Each playable also carries the same ambient ``.play()``; the free function is
the uniform entry that picks the right one.
"""

import inspect

from .base.main import main
from .defs.node import Group, Synth


def play(playable, *, server=None, clock=None, quant=None, controls=None):
    """Play ``playable`` against the ambient context.

    Args:
        playable: an `clausters.seq.event.Event` or a plain dict of event
            keys; an event `clausters.seq.pattern.Pattern`; a
            `clausters.base.stream.Routine` / `clausters.base.stream.Stream`
            or a bare generator (object or function); a bare expression
            (`clausters.defs.Ugen`, `clausters.defs.ChannelList`,
            `clausters.defs.Signal`, `clausters.defs.Box`) or a def
            (`clausters.defs.SynthDef` /
            `clausters.defs.FaustDef` / `clausters.defs.GraphDef`); a
            `clausters.seq.timeline.Timeline`; a `clausters.defs.Buffer`
            (sounded through the stock playbuf instrument); an
            `clausters.seq.automation.Automation` (prepared and triggered);
            or anything with a ``play(destination)`` (the timeline-item
            protocol).
        server: the destination server; ``None`` resolves the ambient one (the
            running session's, else the booted default — see
            `clausters.base.main.Main.resolve_server`).
        clock: the clock to schedule on (patterns, routines and timelines);
            ``None`` resolves the running routine's clock, else the default
            session's (started on first use). Ignored by a bare event played
            immediately and by a def or expression.
        quant: start quantization for a pattern/routine/timeline (see
            `clausters.base.clock.TempoClock.play`).
        controls: ``{name: value}`` controls (ports, for a `GraphDef`) a def
            or expression is instanced with; for a `Buffer`, the stock
            instrument's (``rate``, a musical ratio, and ``amp``). Ignored by
            the other kinds.

    Returns:
        Something that knows how to end what just started: the completed
        event for an event or dict (``.free()`` / ``.release()``), the
        `clausters.seq.eventstream.EventStreamPlayer` for a pattern
        (``.stop()``), the routine for a routine, the node handle — a
        `clausters.defs.Synth` or instance `clausters.defs.Group` — for a
        def, expression or buffer (``.free()``), the
        `clausters.seq.timeline.Playhead` for a timeline (``.stop()``), and
        the `clausters.seq.automation.Automation` itself (``.stop()``).
    """
    from .seq.automation import Automation
    from .seq.event import Event
    from .seq.pattern import Pattern
    from .seq.timeline import Playhead, Timeline
    from .base.stream import Stream, Routine
    from .defs import Buffer, Expr, FaustDef, GraphDef, SynthDef
    from .defs.asdef import as_def

    if isinstance(playable, Event):
        return playable.play(main.resolve_server(server))
    if isinstance(playable, dict):
        return Event(playable).play(main.resolve_server(server))
    if isinstance(playable, (Routine, Stream)):
        clock = clock or main.resolve_clock() or main.get_default_clock()
        return playable.play(clock, quant)
    if inspect.isgenerator(playable) or inspect.isgeneratorfunction(playable):
        routine = _as_routine(playable)
        clock = clock or main.resolve_clock() or main.get_default_clock()
        return routine.play(clock, quant)
    if isinstance(playable, Pattern):
        return playable.play(clock, main.resolve_server(server), quant)
    if isinstance(playable, (Expr, SynthDef, FaustDef, GraphDef)):
        return _play_def(as_def(playable), main.resolve_server(server), controls)
    if isinstance(playable, Timeline):
        clock = clock or main.resolve_clock() or main.get_default_clock()
        playhead = Playhead(playable, clock, main.resolve_server(server))
        playhead.play(quant=quant)
        return playhead
    if isinstance(playable, Buffer):
        return _play_buffer(playable, main.resolve_server(server), controls)
    if isinstance(playable, Automation):
        resolved = main.resolve_server(server)
        if playable.buf is None or playable.bus is None:
            # Interactive trigger: we are off the clock thread, so preparing
            # (allocating and filling the control buffer) may block here.
            playable.prepare(resolved)
        playable.play(resolved)
        return playable
    from .form.element import Element

    if isinstance(playable, Element):
        # An Element carries a timeline-item play() (the hook flattening
        # uses), but the verb keeps the state split: rendering is its door.
        raise TypeError(
            "an arrangement Element is rendered, not played — see "
            "clausters.render / clausters.form.render"
        )
    if callable(getattr(playable, "play", None)):
        # The timeline-item protocol (`OscEvent`, `MidiEvent`, and anything
        # else a Playhead could play): play(destination).
        return playable.play(main.resolve_server(server))
    raise TypeError(
        f"don't know how to play {type(playable).__name__}; expected an Event "
        "or event dict, an event Pattern (Pbind), a Routine/Stream or "
        "generator, a def or bare expression (Ugen/ChannelList/Signal/Box), "
        "a Timeline, "
        "a Buffer, an Automation, or anything with play(destination). An "
        "arrangement Element is rendered, not played — see "
        "clausters.form.render."
    )


def _as_routine(playable):
    """A `Routine` over a generator: a genfunc is wrapped directly; an
    already-created generator object is played through once (a `reset`
    cannot restart it — pass the function to keep it re-runnable)."""
    from .base.stream import Routine

    if inspect.isgeneratorfunction(playable):
        return Routine(playable)
    return Routine(lambda _=None: playable)


def _play_def(d, server, controls):
    """Sends ``d`` (any family) and instances it: ``/graph_new`` for a
    `GraphDef`, ``/synth_new`` otherwise. Returns the node handle."""
    from .defs import GraphDef

    d.send(server)
    if isinstance(d, GraphDef):
        return Group.graph(d.name, controls, server=server)
    return Synth.new(d.name, controls, server=server)


def _play_buffer(buffer, server, controls):
    """A buffer sounds through an instrument (see docs/decisions.md); here
    the verb provides the stock one — one `play_buf` lane per channel, with
    ``rate`` and ``amp`` controls — and frees it when the take ends (the
    buffer's frames over its rate). Returns the `Synth`."""
    from .defs import (
        SynthDef, buf_sample_rate, control, out, play_buf, sample_rate,
    )

    if not buffer.frames and getattr(server.interface, "time_mode",
                                     "unix") != "score":
        buffer.info()      # fills frames/channels/sample_rate
    if not buffer.frames:
        raise ValueError(
            "cannot play a buffer of unknown length; fill the handle's "
            "frames (RT queries the server; NRT needs them up front)")
    channels = max(1, buffer.channels)
    buf = control("buf", 0.0)
    # `rate` is a musical ratio: the def rescales it by the buffer's own
    # sample rate (like sclang's BufRateScale), so a take plays at pitch on
    # any engine rate.
    rate = control("rate", 1.0) * buf_sample_rate(buf) / sample_rate()
    amp = control("amp", 1.0)
    sdef = SynthDef(f"_playbuf{channels}",
                    *[out(float(ch), play_buf(buf, float(ch), rate) * amp)
                      for ch in range(channels)])
    sdef.send(server)
    controls = {"buf": float(buffer.bufnum), **(controls or {})}
    node = Synth.new(sdef.name, controls, server=server)
    file_sr = buffer.sample_rate or 48_000.0
    dur = buffer.frames / file_sr / float(controls.get("rate", 1.0))
    server.send_bundle_after(dur, ("/node_free", node.id))
    return node

File: clients/web/src/play.ts
// The free-standing `play` — one verb for everything playable (mirrors
// `clausters/play.py`).
//
// `play` is the interactive front door: it plays whatever you hand it against
// the ambient context, so you never spell out a server or a clock for a quick
// take. Like SuperCollider's `play` (and the Python client's), it dispatches
// by kind:
//
// - an `Event` — or a plain **object** of event keys — → a note (immediate
//   outside a clock, timetagged inside one);
// - an event `Pattern` (a `Pbind`) → an `EventStreamPlayer` on a clock;
// - a `Routine`/`Stream`, or a bare **generator** (object or function) →
//   scheduled on a clock;
// - a **def** (`SynthDef` / `FaustDef` / `GraphDef`) → sent and instanced on
//   the server. Returns the node handle — it plays until you free it;
// - a bare **expression** (a UGen graph, a `ChannelList`, a Faust `Signal`) →
//   the same, through the ephemeral-def coercion (`defs/asdef.ts`), so
//   `play(sine(440).mul(0.5))` sounds a def it wrapped for you;
// - a `Timeline` → a `Playhead` over the ambient clock and server;
// - an `Automation` → its lane synth triggered and its targets mapped
//   (`await auto.prepare(server)` first — see below);
// - a `Buffer` → sounded through the stock playbuf instrument (a buffer
//   sounds through an instrument; here the verb provides the default one —
//   `rate`/`amp` controls, freed when the take ends);
// - anything else following the **timeline-item protocol**
//   (`play(destination)` — an `OscEvent`, …) → dispatched to it with the
//   ambient server.
//
// Everything resolves against the ambient environment (the running session,
// else the default one): the server defaults to the ambient session's and the
// clock to the running routine's or, outside one, the default session's clock,
// created and started on first use.
//
// ```ts
// const s = (await Session.embed()).adoptDefault();
// play(new Event({ degree: 0 }));                          // one note, now
// play(new Pbind({ degree: new Pseq([0, 2, 4]), dur: 0.5 })); // a phrase
// ```
//
// **One difference.** The reference verb prepares an
// unprepared `Automation` on the spot, blocking off the clock thread; a
// synchronous verb in a page cannot, so an unprepared one is refused by name
// and `await auto.prepare(server)` is the door.

import { main } from "./base/main.ts";
import { Routine, Stream } from "./base/stream.ts";
import type { TempoClock } from "./base/clock.ts";
import { asDef, isExpr } from "./defs/asdef.ts";
import type { Expr } from "./defs/asdef.ts";
import { Buffer } from "./defs/buffer.ts";
import { FaustDef } from "./defs/faustdef.ts";
import { GraphDef } from "./defs/graphdef.ts";
import { Group, Synth } from "./defs/node.ts";
import type { Controls, Node } from "./defs/node.ts";
import type { Server } from "./defs/server/index.ts";
import { SynthDef } from "./defs/synthdef.ts";
import { bufSampleRate, control, out, playBuf, sampleRate } from "./defs/ugens/index.ts";
import { Event } from "./seq/event.ts";
import type { EventDestination } from "./seq/event.ts";
import type { EventStreamPlayer } from "./seq/eventstream.ts";
import { Pattern } from "./seq/pattern.ts";
import { Automation } from "./seq/automation.ts";
import { Playhead, Timeline } from "./seq/timeline.ts";
import type { PlayDestination } from "./seq/timeline.ts";

/** Anything `play` knows how to start. */
export type Playable =
    | Event
    | Record<string, unknown>
    | Pattern<unknown>
    | Stream
    | Generator<number | undefined, unknown, unknown>
    | (() => Generator<number | undefined, unknown, unknown>)
    | SynthDef
    | FaustDef
    | GraphDef
    | Expr
    | Timeline
    | Buffer
    | Automation
    | { play(destination: unknown): unknown };

export interface PlayOptions {
    /** The destination server; the ambient one by default. */
    server?: Server;
    /**
     * The clock to schedule on (patterns, routines, timelines); the running
     * routine's by default, else the default session's, started on first use.
     * Ignored by a bare event played immediately, and by a def or a buffer.
     */
    clock?: TempoClock;
    /** Start quantization for a pattern, routine or timeline. */
    quant?: number;
    /**
     * The controls (ports, for a `GraphDef`) a def is instanced with; for a
     * `Buffer`, the stock instrument's `rate` (a musical ratio) and `amp`.
     */
    controls?: Controls;
}

/**
 * Plays `playable` against the ambient context.
 *
 * Returns something that knows how to end what just started: the completed
 * event for an event or object (`free()` / `release()`), the
 * `EventStreamPlayer` for a pattern (`stop()`), the routine for a routine, the
 * node handle for a def or a buffer (`free()`), the `Playhead` for a timeline
 * (`stop()`) and the `Automation` itself (`stop()`).
 */
export function play(playable: Playable, options: PlayOptions = {}): unknown {
    const { server, clock, quant, controls } = options;

    if (playable instanceof Event) {
        return playable.play(destinationFor(server));
    }
    if (playable instanceof Stream) {
        return playable.play(clock ?? ambientClock(), quant);
    }
    if (playable instanceof Pattern) {
        return playable.play(destinationFor(server), { clock, quant }) as EventStreamPlayer;
    }
    if (isGenerator(playable) || isGeneratorFunction(playable)) {
        return asRoutine(playable).play(clock ?? ambientClock(), quant);
    }
    if (
        playable instanceof SynthDef
        || playable instanceof FaustDef
        || playable instanceof GraphDef
    ) {
        return playDef(playable, main.resolveServer(server), controls);
    }
    if (isExpr(playable)) {
        // A bare expression: coerced into an ephemeral def and played like
        // any other. The def is named `tmp_*`, so the server never persists
        // it, and it is this call's alone.
        return playDef(asDef(playable), main.resolveServer(server), controls);
    }
    if (playable instanceof Timeline) {
        const playhead = new Playhead(
            playable,
            clock ?? ambientClock(),
            main.resolveServer(server) as unknown as PlayDestination,
        );
        playhead.play({ quant });
        return playhead;
    }
    if (playable instanceof Buffer) {
        return playBuffer(playable, main.resolveServer(server), controls);
    }
    if (playable instanceof Automation) {
        playable.play(main.resolveServer(server));
        return playable;
    }
    if (typeof playable === "object" && typeof (playable as Playable & {
        play?: unknown;
    }).play === "function") {
        // The timeline-item protocol (`OscEvent`, and anything else a
        // Playhead could play): play(destination).
        return (playable as { play(destination: unknown): unknown })
            .play(main.resolveServer(server));
    }
    if (typeof playable === "object" && playable !== null) {
        return new Event(playable as Record<string, unknown>).play(destinationFor(server));
    }
    throw new TypeError(
        `don't know how to play ${String(playable)}; expected an Event or event ` +
            "object, an event Pattern (Pbind), a Routine/Stream or generator, a " +
            "def (SynthDef/FaustDef/GraphDef) or a bare expression, a Timeline, a " +
            "Buffer, an " +
            "Automation, or anything with play(destination)",
    );
}

/** The ambient server as an event destination. */
function destinationFor(server?: Server): EventDestination {
    return main.resolveServer(server) as unknown as EventDestination;
}

/** The clock an ambient play schedules on, created and started on first use. */
function ambientClock(): TempoClock {
    return main.resolveClock() ?? main.getDefaultClock();
}

function isGenerator(value: unknown): value is Generator<number | undefined, unknown, unknown> {
    return (
        typeof value === "object" && value !== null
        && typeof (value as { next?: unknown }).next === "function"
        && typeof (value as { [Symbol.iterator]?: unknown })[Symbol.iterator] === "function"
    );
}

function isGeneratorFunction(value: unknown): boolean {
    return typeof value === "function"
        && value.constructor?.name === "GeneratorFunction";
}

/**
 * A `Routine` over a generator: a generator *function* is wrapped directly; an
 * already-created generator object is played through once (a `reset` cannot
 * restart it — pass the function to keep it re-runnable).
 */
function asRoutine(playable: unknown): Routine {
    if (isGeneratorFunction(playable)) {
        return new Routine(playable as () => Generator<number | undefined, unknown, unknown>);
    }
    return new Routine(() => playable as Generator<number | undefined, unknown, unknown>);
}

/**
 * Sends `def` (any family) and instances it: `/graph_new` for a `GraphDef`,
 * `/synth_new` otherwise. Returns the node handle.
 *
 * The send is not awaited — `play` is a synchronous verb — but the carrier is
 * ordered, so the def is on its way before the creation that names it.
 */
function playDef(
    def: SynthDef | FaustDef | GraphDef,
    server: Server,
    controls?: Controls,
): Node {
    void def.send(server, { wait: false });
    if (def instanceof GraphDef) {
        return Group.graph(def.name, controls, { server });
    }
    return new Synth(def.name, controls, { server });
}

/**
 * A buffer sounds through an instrument (see `docs/decisions.md`); here the
 * verb provides the stock one — one `playBuf` lane per channel, with `rate`
 * and `amp` controls — and frees it when the take ends (the buffer's frames
 * over its rate). Returns the `Synth`.
 */
function playBuffer(buffer: Buffer, server: Server, controls?: Controls): Synth {
    if (!buffer.frames) {
        throw new Error(
            "cannot play a buffer of unknown length; read its shape first "
                + "(await buffer.info())",
        );
    }
    const channels = Math.max(1, buffer.channels);
    const buf = control("buf", 0.0);
    // `rate` is a musical ratio: the def rescales it by the buffer's own
    // sample rate, so a take plays at pitch on any engine rate.
    const rate = control("rate", 1.0).mul(bufSampleRate(buf)).div(sampleRate());
    const amp = control("amp", 1.0);
    const def = new SynthDef(
        `_playbuf${channels}`,
        ...Array.from({ length: channels }, (_, ch) =>
            out(ch, playBuf(buf, ch, rate).mul(amp))),
    );
    void def.send(server, { wait: false });
    const given = controls ? Object.fromEntries(Object.entries(controls)) : {};
    const node = new Synth(def.name, { buf: buffer.bufnum, ...given }, { server });
    const fileRate = buffer.sampleRate || 48_000.0;
    const speed = Number((given as Record<string, number>).rate ?? 1.0);
    server.sendBundleAfter(buffer.frames / fileRate / speed, [
        ["/node_free", ["i", node.id]],
    ]);
    return node;
}
